src/TodoApp.tsx

```diff
@@ -0,0 +1,158 @@
+import React, {
+  FormEvent,
+  useEffect,
+  useState,
+} from 'react';
+import { useLocation } from 'react-router-dom';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
+import { Footer } from './components/Footer';
+import { FilterTypes } from './types/FilterTypes';
+
+export const TodoApp: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>(() => {
+    const localTodos = localStorage.getItem('todos');
+
+    try {
+      return localTodos ? JSON.parse(localTodos) : [];
+    } catch {
+      return [];
+    }
+  });
+
+  const [title, setTitle] = useState('');
+  const { pathname } = useLocation();
+
+  useEffect(() => {
+    localStorage.setItem('todos', JSON.stringify(todos));
+  }, [todos]);
+
+  const filteredTodos = todos.filter(({ completed }) => {
+    switch (pathname) {
+      case FilterTypes.COMPLETED:
+        return completed;
+
+      case FilterTypes.ACTIVE:
+        return !completed;
+
+      default:
+        return true;
+    }
+  });
+
+  const addNewTodo = (event: FormEvent) => {
+    event.preventDefault();
+    const newTitle = title.trim();
+
+    if (!newTitle) {
+      return;
+    }
+
+    const newId = +new Date();
+
+    const todoToAdd = {
+      id: newId,
+      title: newTitle,
+      completed: false,
+    };
+
+    setTodos(previousTodos => [...previousTodos, todoToAdd]);
+    setTitle('');
+  };
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+
+  const switchTodoStatus = (id: number, status: boolean) => {
+    setTodos(
+      todos.map((todo: Todo) => {
+        return {
+          ...todo,
+          ...(id === todo.id && { completed: status }),
+        };
+      }),
+    );
+  };
+
+  const toggleStatus = (
+    id: number,
+    status: boolean,
+  ) => {
+    switchTodoStatus(id, !status);
+  };
+
+  const toggleAllTodos = () => {
+    const someTodoActive = todos.some(todo => !todo.completed);
+
+    setTodos(todos.map(todo => ({
+      ...todo,
+      completed: Boolean(someTodoActive),
+    })));
+  };
+
+  const deleteTodo = async (id: number) => {
+    setTodos(todos.filter((todo: Todo) => id !== todo.id));
+  };
+
+  const removeCompletedTodos = () => {
+    setTodos(todos.filter(todo => !todo.completed));
+  };
+
+  const changeInputText = (id: number, query: string) => {
+    setTodos(todos.map(todo => {
+      if (id === todo.id) {
+        return {
+          ...todo,
+          title: query,
+        };
+      }
+
+      return todo;
```

Comment:
understanding such style will help you in the future with database queries or other form attachments, but it's up to you

```suggestion
        return {
          ...todo,
          ...(id === todo.id && { title: query }),
        };
```